# Exercises: Evaluation

<hr style="border:2px solid gray">

![](confusion_matrix.png)

In [1]:
import pandas as pd

1. Create a new file named model_evaluation.py or model_evaluation.ipynb for these exercises.

2. Given the following confusion matrix, evaluate (by hand) the model's performance.

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |


- In the context of this problem, what is a false positive?
- In the context of this problem, what is a false negative?
- How would you describe this model?

3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.
- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:
- a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?
- b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?
- c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?    

5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

- sklearn.metrics.accuracy_score
- sklearn.metrics.precision_score
- sklearn.metrics.recall_score
- sklearn.metrics.classification_report

<hr style="border:2px solid black">

<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
Real life example of False Positive: COVID test says you are positive, but you actually negative
<br>
    <br>
Real life example of False Negative: COVID test says you are negative, but you actually positive!
</div>

<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
<b>When do we use accuracy/precision/recall?</b>
    <br>
1. We can use accuracy when we are interested in predicting both 0 and 1 correctly and our dataset is balanced enough. 
    <br>
2. We use precision when we want the prediction of 1 to be as correct as possible.
    <br>
3. We use recall when we want our model to spot as many real 1 as possible. 
    <br>
    <br>
    <b>    Goal of precision and recall</b>: To avoiding as many mistakes as possible in your model
<br>
<br>
<b>Example</b>: You have a classification model that predicts apples and bananas
    <br>
-- <b>Recall</b>: avoids predicting apples as bananas
    <br>
--- Use cases: in high stakes situations
    <br>
---- ie: Finding cancer in a patient. It would be better to say there was a mistake in testing and that patient doesn't actually have cancer(False Positive) than to say they didn't have cancer but they actually do! (False Negative)    
        <br>
        <br>
-- <b>Precision</b>: avoids mistakes in predicting bananas as apples.
    <br>
    <br>    
    -- <b>F1</b>: takes both precision <b>and</b> recall into account.  

</div>


### #2. Given the following confusion matrix, evaluate (by hand) the model's performance.

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |


- cat = positive class

- dog = negative class

In the context of this problem, what is a false positive?

- False positive: we predicted a cat, but its a dog

In the context of this problem, what is a false negative?

- False negative: We predicted a dog, but its a cat

How would you describe this model?

In [2]:
 #true positive is predicting its a cat, and its a cat
tp = 34

#true negative is predicting its a dog, and its a dog
tn = 46

#false positive is predicting its a cat, but its a dog
fp = 7

#false negative is predicting its a dog, but its a cat
fn = 13

In [3]:
print("Cat-classifier (where 'cat' is the positive prediction)")

print("True Positives:", tp)
print("False Positives:", fp)
print("False Negatives:", fn)
print("True Negatives:", tn)

print("-------------")

accuracy = (tp + tn) / (tp + tn + fp + fn)

precision = tp / (tp + fp)

recall = tp / (tp + fn)

print("Accuracy is:", accuracy)
print("Recall is:", round(recall,2))
print("Precision is:", round(precision,2))

Cat-classifier (where 'cat' is the positive prediction)
True Positives: 34
False Positives: 7
False Negatives: 13
True Negatives: 46
-------------
Accuracy is: 0.8
Recall is: 0.72
Precision is: 0.83


<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
    <u>recall</u> is about real positives. 
<br>
    <u>precision</u> is about predictive positives.
</div>

<hr style="border:1.5px solid black">

### #3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible.  

<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
Be sure that students have access to the 'c3.csv' file found in the curriculum to do this exercises. 
</div>

In [4]:
#bring in the csv provided in the curriculum exercises
cody_df = pd.read_csv('~/Downloads/c3.csv')

In [5]:
#take a look at the data
cody_df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [6]:
#what kind of columns and dtypes are we dealing with?
cody_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   actual  200 non-null    object
 1   model1  200 non-null    object
 2   model2  200 non-null    object
 3   model3  200 non-null    object
dtypes: object(4)
memory usage: 6.4+ KB


In [7]:
#how many defects and non-defects do we have in the actual data?
cody_df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

- Which evaluation metric would be appropriate here?

<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
Evaluation Metric options are:<br>
<br>
<br>
-- Accuracy: the number of times we predicted correctly divided by the total number of observations. 
<br>
<br>
-- Precision: the percentage of positive predictions that we made that are correct.
<br>
<br>
-- Recall(Sensativity): the percentage of positive cases that we accurately predicted.
    <br>
    <br>
-- Specificity: the percentage of negative cases that we accurately predicted.
    tn/(tn+fp)
</div>

Since we are interested in 'defects', we will asssign it as 'positive class' for the classifier.
<br>
- defects = positive class
<br>

Quality Control, our internal customer, wants the metric to identify as many defective ducks as possible
<br>

Our best metric for Quality Control here is <b>recall</b>
<br>

- i.e how many real positives do we have?
- i.e how many of defective ducks are actually flagged by defective (positive) by the models?
- i.e let's minimize our false negatives


<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
Recall is about real positives
<br>
<br>
recall = tp/(tp+fn)
<br>
<br>    
    false negative is when we say it's not defective, but it is 
</div>

In [8]:
# Model positives
subset = cody_df [cody_df.actual == 'Defect']
subset

actual     model1     model2     model3
13   Defect  No Defect     Defect     Defect
30   Defect     Defect  No Defect     Defect
65   Defect     Defect     Defect     Defect
70   Defect     Defect     Defect     Defect
74   Defect  No Defect  No Defect     Defect
87   Defect  No Defect     Defect     Defect
118  Defect  No Defect     Defect  No Defect
135  Defect     Defect  No Defect     Defect
140  Defect  No Defect     Defect     Defect
147  Defect     Defect  No Defect     Defect
163  Defect     Defect     Defect     Defect
171  Defect  No Defect     Defect     Defect
176  Defect  No Defect     Defect     Defect
186  Defect  No Defect  No Defect  No Defect
194  Defect     Defect  No Defect     Defect
196  Defect     Defect  No Defect  No Defect

- Which model would be the best fit for this use case?

In [9]:
(subset.actual == subset.model1).mean()

0.5

In [10]:
#Model 1 recall
model_recall = (subset.actual == subset.model1).mean()

print("Model 1")
print(f"Model recall: {model_recall:.2%}")

Model 1
Model recall: 50.00%


In [11]:
# Model 2 recall
model_recall = (subset.actual == subset.model2).mean()

print("Model 2")
print(f"Model recall: {model_recall:.2%}")

Model 2
Model recall: 56.25%


In [12]:
# Model 3 recall
model_recall = (subset.actual == subset.model3).mean()

print("Model 3")
print(f"Model recall: {model_recall:.2%}")

Model 3
Model recall: 81.25%


<div class="alert alert-block alert-success">
    <b>Takeaways:</b>
        <br>
        <br>
-- Quality Control should select a model with higher recall (to avoid false negatives)
<br>
    -- Quality Control should use <b>Model 3</b>
</div>

Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect.  

Which evaluation metric would be appropriate here?

- positive case = defect

<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
false positive is when we say its defective, but it is not
<br>
<br>
false negative is when we say not defective, but it is
<br>
<br>    
precision = tp / (tp + fp)
</div>

<div class="alert alert-block alert-success">
    <b>Takeaways:</b>
        <br>
        <br>
PR team really wants to minimize the False positives - meaning choose model with highest precision.
</div>

Which model would be the best fit for this use case?

In [13]:
#precision is about positive predictions 

In [14]:
# choose subset of model1 where we only select 'positive predictions'
subset = cody_df [cody_df.model1 == 'Defect']
subset

actual  model1     model2     model3
3    No Defect  Defect     Defect     Defect
30      Defect  Defect  No Defect     Defect
62   No Defect  Defect  No Defect  No Defect
65      Defect  Defect     Defect     Defect
70      Defect  Defect     Defect     Defect
135     Defect  Defect  No Defect     Defect
147     Defect  Defect  No Defect     Defect
163     Defect  Defect     Defect     Defect
194     Defect  Defect  No Defect     Defect
196     Defect  Defect  No Defect  No Defect

In [15]:
# calculate precision
model_precision = (subset.actual == subset.model1).mean()

print("Model 1")
print(f"Model precision: {model_precision:.2%}")

Model 1
Model precision: 80.00%


In [16]:
# choose subset for model2 where we only select 'positive predictions'
subset = cody_df [cody_df.model2 == 'Defect']

# calculate precision
model_precision = (subset.actual == subset.model2).mean()

print("Model 2")
print(f"Model precision: {model_precision:.2%}")

Model 2
Model precision: 10.00%


In [17]:
# choose subset for model3 where we only select 'positive predictions'
subset = cody_df [cody_df.model3 == 'Defect']

# calculate precision
model_precision = (subset.actual == subset.model3).mean()

print("Model 3")
print(f"Model precision: {model_precision:.2%}")

Model 3
Model precision: 13.13%


<div class="alert alert-block alert-success">
    <b>Takeaways:</b>
        <br>
        <br>
        Use model 1 since it will minimize the false positive predictions of defects
        </div>

<hr style="border:1.5px solid black">

### #4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).
Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
Be sure that students have access to the gives_you_paws.csv file found in the curriculum to do this exercises. 
    </div>

In [18]:
paws_df = pd.read_csv('gives_you_paws.csv')

In [19]:
#take a look at the data
paws_df.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [20]:
#what kind of columns and dtypes are we dealing with?
paws_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   actual  5000 non-null   object
 1   model1  5000 non-null   object
 2   model2  5000 non-null   object
 3   model3  5000 non-null   object
 4   model4  5000 non-null   object
dtypes: object(5)
memory usage: 195.4+ KB


In [21]:
#what are our actual counts
paws_df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

<div class="alert alert-block alert-success">
    <b>Takeaways:</b>
        <br>
        <br>
It looks like in out actual data collection, there are almost twice as many dog samples
        </div>

In [22]:
#set the most common class ('dog') as the baseline
paws_df['baseline'] = paws_df.actual.value_counts().idxmax()

<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
-- .idxmax() returns the value with the maximum count.
<br>
    -- we can set this as the baseline 'model'
</div>

A. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In [23]:
#baseline accuracy 
(paws_df.actual == paws_df.baseline).mean()

0.6508

In [24]:
#model 1 accuracy
(paws_df.model1 == paws_df.actual).mean()

0.8074

In [25]:
#model 2 accuracy
(paws_df.model2 == paws_df.actual).mean()

0.6304

In [26]:
#model 3 accuracy
(paws_df.model3 == paws_df.actual).mean()

0.5096

In [27]:
#model 4 accuracy
(paws_df.model4 == paws_df.actual).mean()

0.7426

In [28]:
#let's put it all together
model_acc = []

for model in paws_df.columns[1:]:
    acc = (paws_df.actual == paws_df[model]).mean()
    model_acc.append([model, acc])

model_acc

[['model1', 0.8074],
 ['model2', 0.6304],
 ['model3', 0.5096],
 ['model4', 0.7426],
 ['baseline', 0.6508]]

In [29]:
#make pretty in df
pd.DataFrame(model_acc, columns=['model','accuracy'])

model  accuracy
0    model1    0.8074
1    model2    0.6304
2    model3    0.5096
3    model4    0.7426
4  baseline    0.6508

<div class="alert alert-block alert-success">
    <b>Takeaways:</b>
        <br>
        <br>
In terms of accuracy, model 1 and model 3 perform better than baseline
</div>

B. Suppose you are working on a team that solely deals with <b>dog</b> pictures. Which of these models would you recommend?

- dog = positive class
- cat = negative class

<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
-- precision = tp / (tp + fp)
<br>
-- recall = tp / (tp + fn)
<br>
-- false positives are when we say its a dog, but its actually a cat
<br>
-- false negative is when we say its a cat, but its actually a dog
</div>

<b>Phase I</b>: Automated algorithm tags pictures as either a cat or a dog
<br>
For Phase I, we should choose a model with highest Recall

In [32]:
subset = paws_df[paws_df.actual == 'dog']
subset.head()

actual model1 model2 model3 model4 baseline
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
5    dog    dog    dog    dog    dog      dog
8    dog    dog    cat    dog    dog      dog

In [33]:
# Model 1 Recall
(subset.actual == subset.model1).mean()

0.803318992009834

In [34]:
# Model 2 Recall
(subset.actual == subset.model2).mean()

0.49078057775046097

In [35]:
# Model 3 Recall
(subset.actual == subset.model3).mean()

0.5086047940995697

In [36]:
# Model 4 Recall
(subset.actual == subset.model4).mean()

0.9557467732022127

<div class="alert alert-block alert-success">
    <b>Takeaways:</b>
        <br>
        <br>
It appears that Model 4 is performing the best, with Recall of 0.96
</div>

<b>Phase II</b>: Photos that have been initially identified are put through another round of review
<br>
People have a service to see dog pictures, so what do we want to minimize?
- we want to minimize false positives
<br>
Precision is the appropriate metric since we are trying to minimize false positives

In [37]:
#take another look
paws_df.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [38]:
subset1 = paws_df[paws_df.model1 == 'dog']
subset2 = paws_df[paws_df.model2 == 'dog']
subset3 = paws_df[paws_df.model3 == 'dog']
subset4 = paws_df[paws_df.model4 == 'dog']

In [39]:
# Model 1 Precision
(subset1.actual == subset1.model1).mean()

0.8900238338440586

In [40]:
# Model 2 Precision
(subset2.actual == subset2.model2).mean()

0.8931767337807607

In [41]:
# Model 3 Precision
(subset3.actual == subset3.model3).mean()

0.6598883572567783

In [42]:
# Model 4 Precision
(subset4.actual == subset4.model4).mean()

0.7312485304490948

<div class="alert alert-block alert-success">
    <b>Takeaways:</b>
        <br>
        <br>
It appears that Model 2 is performing best with Precision of 0.893
</div>

In [44]:
#Let's visualize this with a crosstab
pd.crosstab(paws_df.model2, paws_df.actual)

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

In [45]:
#Let's put it together with all models
model_pre = []

for model in paws_df.columns[1:]:
    
    subset = paws_df [paws_df[model] == 'dog']
    
    precision = (subset.actual == subset[model]).mean()

    model_pre.append([model,precision])
    
model_pre

[['model1', 0.8900238338440586],
 ['model2', 0.8931767337807607],
 ['model3', 0.6598883572567783],
 ['model4', 0.7312485304490948],
 ['baseline', 0.6508]]

In [46]:
#make pretty in dataframe
pd.DataFrame(model_pre, columns=['model','precision'])

model  precision
0    model1   0.890024
1    model2   0.893177
2    model3   0.659888
3    model4   0.731249
4  baseline   0.650800

<div class="alert alert-block alert-success">
    <b>Takeaway for Dog team:</b>
        <br>
        <br>
-- we had to maximize precision to minimize the false positives
        <br>
-- therefore, we could use model 2 since its the highest
        <br>
-- since model 1 and model 2 are so similar, we could evaluate on an additional metric
</div>

C. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend? 

<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
-- false positive is when we say its a cat, but its a dog
<br>
-- false negative is when say its a dog, but its a cat
</div>

- cat = positive class
- dog = negative class

We want to minimize the false positives, therefore, we will use precision again.

In [47]:
#calculate for all models
model_pre = []

for model in paws_df.columns[1:]:
    
    subset = paws_df [paws_df[model] == 'cat']
    
    precision = (subset.actual == subset[model]).mean()

    model_pre.append([model,precision])
    
model_pre

[['model1', 0.6897721764420747],
 ['model2', 0.4841220423412204],
 ['model3', 0.358346709470305],
 ['model4', 0.8072289156626506],
 ['baseline', nan]]

In [48]:
#make pretty in df
pd.DataFrame(model_pre, columns=['model','precision'])

model  precision
0    model1   0.689772
1    model2   0.484122
2    model3   0.358347
3    model4   0.807229
4  baseline        NaN

<div class="alert alert-block alert-success">
    <b>Takeaway for Cat team:</b>
        <br>
        <br>
-- we had to maximize precision to minimize the false positives
<br>
-- therefore, we should use model 4
        </div>

<hr style="border:1.5px solid black">

### #5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

In [50]:
from sklearn.metrics import classification_report

- sklearn.metrics.accuracy_score
- sklearn.metrics.precision_score
- sklearn.metrics.recall_score
- sklearn.metrics.classification_report

In [51]:
#classification report of model 1
print("Model 1")
pd.DataFrame(classification_report(paws_df.actual, paws_df.model1, 
                      labels=['cat','dog'],
                      output_dict=True)).T

Model 1


precision    recall  f1-score    support
cat            0.689772  0.815006  0.747178  1746.0000
dog            0.890024  0.803319  0.844452  3254.0000
accuracy       0.807400  0.807400  0.807400     0.8074
macro avg      0.789898  0.809162  0.795815  5000.0000
weighted avg   0.820096  0.807400  0.810484  5000.0000

In [52]:
#classification report of model 2
print("Model 2")
pd.DataFrame(classification_report(paws_df.actual, paws_df.model2, 
                      labels=['cat','dog'],
                      output_dict=True)).T

Model 2


precision    recall  f1-score    support
cat            0.484122  0.890607  0.627269  1746.0000
dog            0.893177  0.490781  0.633479  3254.0000
accuracy       0.630400  0.630400  0.630400     0.6304
macro avg      0.688649  0.690694  0.630374  5000.0000
weighted avg   0.750335  0.630400  0.631310  5000.0000

In [53]:
#classification report of model 3
print("Model 3")
pd.DataFrame(classification_report(paws_df.actual, paws_df.model3, 
                      labels=['cat','dog'],
                      output_dict=True)).T

Model 3


precision    recall  f1-score    support
cat            0.358347  0.511455  0.421425  1746.0000
dog            0.659888  0.508605  0.574453  3254.0000
accuracy       0.509600  0.509600  0.509600     0.5096
macro avg      0.509118  0.510030  0.497939  5000.0000
weighted avg   0.554590  0.509600  0.521016  5000.0000

In [54]:
#classification report of model 4
print("Model 4")
pd.DataFrame(classification_report(paws_df.actual, paws_df.model4, 
                      labels=['cat','dog'],
                      output_dict=True)).T

Model 4


precision    recall  f1-score    support
cat            0.807229  0.345361  0.483755  1746.0000
dog            0.731249  0.955747  0.828560  3254.0000
accuracy       0.742600  0.742600  0.742600     0.7426
macro avg      0.769239  0.650554  0.656157  5000.0000
weighted avg   0.757781  0.742600  0.708154  5000.0000

In [55]:
from sklearn.metrics import precision_score, recall_score

<div class="alert alert-block alert-info">
<b>Instructor Note:</b>
<br>
<br>
-- We can create a function to give us precision and recall
</div>

In [60]:
def calculate_precision(predictions, positive='dog'):
    return precision_score(paws_df.actual, predictions, pos_label=positive)

In [61]:
def calculate_recall(predictions, positive='dog'):
    return recall_score(paws_df.actual, predictions, pos_label=positive)

In [62]:
pd.concat([
    paws_df.loc[:, 'model1':'baseline'].apply(calculate_recall).rename('recall'),
    paws_df.loc[:, 'model1':'baseline'].apply(calculate_precision).rename('precision'),
], axis=1)

recall  precision
model1    0.803319   0.890024
model2    0.490781   0.893177
model3    0.508605   0.659888
model4    0.955747   0.731249
baseline  1.000000   0.650800